crates/connector_proxy/src/libs/stream.rs

```diff
@@ -1,25 +1,244 @@
+        let items = in_stream
+            .map(move |bytes| {
+                let b = bytes?;
+                buf.extend_from_slice(b.chunk());
+                let chunk = buf.chunk();
+
+                // Deserialize to Value first, instead of Message, to avoid missing 'is_eof' signals in error.
+                let deserializer = Deserializer::from_slice(chunk);
+                let mut value_stream = deserializer.into_iter::<Value>();
+
+                // Turn Values into Messages and validate them
+                let values: Vec<Result<Message, std::io::Error>> = value_stream
+                    .by_ref()
+                    .map_while(|value| match value {
+                        Ok(v) => Some(Ok(v)),
+                        Err(e) => {
+                            // we must stop as soon as we hit EOF to avoid
+                            // progressing value_stream.byte_offset() so that we can
+                            // safely drop the buffer up to byte_offset() and pick up the leftovers
+                            // when working with the next bytes
+                            if e.is_eof() {
+                                return None;
+                            }
+
+                            Some(raise_err(&format!(
+                                "error in decoding JSON: {:?}, {:?}",
+                                e,
+                                std::str::from_utf8(chunk)
+                            )))
+                        }
+                    })
+                    .map(|value| match value {
+                        Ok(v) => {
+                            let message: Message = match serde_json::from_value(v) {
+                                Ok(m) => m,
+                                // We ignore JSONs that are not Airbyte Messages according
+                                // to the specification:
+                                // https://docs.airbyte.com/understanding-airbyte/airbyte-specification#the-airbyte-protocol
+                                Err(_) => return Ok(None),
+                            };
+
+                            message.validate().map_err(|e| {
+                                create_custom_error(&format!("error in validating message {:?}", e))
+                            })?;
+
+                            tracing::debug!("read message:: {:?}", &message);
+                            Ok(Some(message))
+                        }
+                        Err(e) => Err(e),
+                    })
+                    // Flipping the Option and Result to filter out the None values
+                    .filter_map(|value| match value {
+                        Ok(Some(v)) => Some(Ok(v)),
+                        Ok(None) => None,
+                        Err(e) => Some(Err(e)),
+                    })
+                    .collect();
+
+                let byte_offset = value_stream.byte_offset();
+                drop(buf.split_to(byte_offset));
```

Comment:
@psFried Hmm, how would we be able to verify it this is actually freeing up memory? I'm also not sure if this is doing anything and if it's necessary at all.

Reply:
Based on my reading of the [`BytesMut` docs](https://docs.rs/bytes/latest/bytes/struct.BytesMut.html), particularly this bit:

> BytesMut can be thought of as containing a buf: Arc<Vec<u8>>

I think this indicates that we can't rely on `split_to` to actually free anything. For now, it might be reasonable to use [`Chain`](https://docs.rs/bytes/latest/bytes/buf/struct.Chain.html) type to create a continuous view into the underlying sequence of `Bytes` values from `in_stream`, rather than copying everything into a single shared buffer.

Reply:
`BytesMut` does have an implementation for `Drop` trait: https://docs.rs/bytes/latest/src/bytes/bytes_mut.rs.html#925-940

And it seems to handle the case of `Arc<Vec>` as well. WDYT?

Reply:
Given the new implementation, we don't need a buffer at all anymore. See the other comment below and let me know what you think.